tutorials/multi/2_node_classification.py

```diff
@@ -0,0 +1,249 @@
+    
+    # Define training and validation dataloader, copied from the previous tutorial
+    # but with one line of difference: use_ddp to enable distributed data parallel
+    # data loading.
+    sampler = dgl.dataloading.MultiLayerNeighborSampler([4, 4])
+    train_dataloader = dgl.dataloading.NodeDataLoader(
+        # The following arguments are specific to NodeDataLoader.
+        graph,              # The graph
+        train_nids,         # The node IDs to iterate over in minibatches
+        sampler,            # The neighbor sampler
+        device=device,      # Put the sampled MFGs on CPU or GPU
+        use_ddp=True,       # Make it work with distributed data parallel
+        # The following arguments are inherited from PyTorch DataLoader.
+        batch_size=1024,    # Per-device batch size.
+                            # The effective batch size is this number times the number of GPUs.
+        shuffle=True,       # Whether to shuffle the nodes for every epoch
+        drop_last=False,    # Whether to drop the last incomplete batch
+        num_workers=0       # Number of sampler processes
+    )
+    valid_dataloader = dgl.dataloading.NodeDataLoader(
+        graph, valid_nids, sampler,
+        batch_size=1024,
+        shuffle=False,
+        drop_last=False,
+        num_workers=0,
+        device=device,
+        use_ddp=False
+    )
+    
+    model = Model(num_features, 128, num_classes).to(device)
+    # Wrap the model with distributed data parallel module.
+    if device == torch.device('cpu'):
+        model = torch.nn.parallel.DistributedDataParallel(model, device_ids=None, output_device=None)
+    else:
+        model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[device], output_device=device)
+    
+    # Define optimizer
+    opt = torch.optim.Adam(model.parameters())
+    
+    best_accuracy = 0
+    best_model_path = './model.pt'
+    
+    # Copied from previous tutorial with changes highlighted.
+    for epoch in range(10):
+        train_dataloader.set_epoch(epoch)    # <--- necessary for dataloader with DDP.
+        model.train()
+
+        with tqdm.tqdm(train_dataloader) as tq:
+            for step, (input_nodes, output_nodes, mfgs) in enumerate(tq):
+                # feature copy from CPU to GPU takes place here
+                inputs = mfgs[0].srcdata['feat']
+                labels = mfgs[-1].dstdata['label']
+
+                predictions = model(mfgs, inputs)
+
+                loss = F.cross_entropy(predictions, labels)
+                opt.zero_grad()
+                loss.backward()
+                opt.step()
+
+                accuracy = sklearn.metrics.accuracy_score(labels.cpu().numpy(), predictions.argmax(1).detach().cpu().numpy())
+
+                tq.set_postfix({'loss': '%.03f' % loss.item(), 'acc': '%.03f' % accuracy}, refresh=False)
+
+        model.eval()
+
+        # Evaluate on only the first GPU.
+        if proc_id == 0:
```

Comment:
@BarclayII I've noticed in the other DGL examples, validation/testing is often done on just one GPU. What's the reason for this? In the multi-GPU cases, it's generally slower than the training due to this.

Reply:
I'm not sure what is an elegant way to perform multi-GPU validation and testing because I will need to send the results to a single process where the final validation/test accuracy is computed.  I remember RGCN has multi-GPU validation but the code is quite tedious (it has a queue and something).